services/s3/src/main/java/software/amazon/awssdk/services/s3/S3Utilities.java

```diff
@@ -251,6 +257,168 @@ public URL getUrl(GetUrlRequest getUrlRequest) {
         }
     }
 
+    /**
+     * Returns a parsed {@link S3Uri} with which a user can easily retrieve the bucket, key, region, style, and query
+     * parameters of the URI. Only path-style and virtual-hosted-style URI parsing is supported, including CLI-style
+     * URIs, e.g., "s3://bucket/key". AccessPoints and Outposts URI parsing is not supported. If you work with object keys
+     * and/or query parameters with special characters, they must be URL-encoded, e.g., replace " " with "%20". If you work with
+     * bucket names that contain a dot, i.e., ".", the dot must not be URL-encoded. Encoded buckets, keys, and query parameters
+     * will be returned decoded.
+     *
+     * <p>
+     * For more information on path-style and virtual-hosted-style URIs, see <a href=
+     * "https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-bucket-intro.html"
+     * >Methods for accessing a bucket</a>.
+     *
+     * @param uri The URI to be parsed
+     * @return Parsed {@link S3Uri}
+     *
+     * <p><b>Example Usage</b>
+     * <p>
+     * {@snippet :
+     *     S3Client s3Client = S3Client.create();
+     *     S3Utilities s3Utilities = s3Client.utilities();
+     *     String uriString = "https://myBucket.s3.us-west-1.amazonaws.com/doc.txt?versionId=abc123";
+     *     URI uri = URI.create(uriString);
+     *     S3Uri s3Uri = s3Utilities.parseUri(uri);
+     *
+     *     String bucket = s3Uri.bucket().get(); // "myBucket"
```

Comment:
We probably shouldn't just do .get on an Optional since it's a bad practice.

Maybe `s3Uri.bucket.ifPresent(bucket -> System.out.println("bucket: " + bucket)); // "myBucket"`? Same for other properties below